specs/phase1/shard-fork-choice.md

```diff
@@ -0,0 +1,138 @@
+    return ShardStore(
+        shard=shard,
+        blocks={anchor_state.shard_states[shard].latest_block_root: ShardBlock(slot=anchor_state.slot)},
+        block_states={anchor_state.shard_states[shard].latest_block_root: anchor_state.copy().shard_states[shard]},
+    )
+```
+
+#### `get_shard_latest_attesting_balance`
+
+```python
+def get_shard_latest_attesting_balance(store: Store, shard_store: ShardStore, root: Root) -> Gwei:
+    state = store.checkpoint_states[store.justified_checkpoint]
+    active_indices = get_active_validator_indices(state, get_current_epoch(state))
+    return Gwei(sum(
+        state.validators[i].effective_balance for i in active_indices
+        if (
+            i in store.latest_messages and get_shard_ancestor(
+                store, shard_store, store.latest_messages[i].root, shard_store.blocks[root].slot
+            ) == root
+        )
+    ))
+```
+
+#### `get_shard_head`
+
+```python
+def get_shard_head(store: Store, shard_store: ShardStore) -> Root:
+    # Execute the LMD-GHOST fork choice
+    head_beacon_root = get_head(store)
+    head_shard_root = store.block_states[head_beacon_root].shard_states[shard_store.shard].latest_block_root
+    while True:
+        children = [
+            root for root in shard_store.blocks.keys()
+            if shard_store.blocks[root].shard_parent_root == head_shard_root
+        ]
+        if len(children) == 0:
+            return head_shard_root
+        # Sort by latest attesting balance with ties broken lexicographically
+        head_shard_root = max(
+            children, key=lambda root: (get_shard_latest_attesting_balance(store, shard_store, root), root)
+        )
+```
+
+#### `get_shard_ancestor`
+
+```python
+def get_shard_ancestor(store: Store, shard_store: ShardStore, root: Root, slot: Slot) -> Root:
+    block = shard_store.blocks[root]
+    if block.slot > slot:
+        return get_shard_ancestor(store, shard_store, block.shard_parent_root, slot)
+    elif block.slot == slot:
+        return root
+    else:
+        # root is older than queried slot, thus a skip slot. Return earliest root prior to slot
+        return root
+```
+
+### Handlers
+
+#### `on_shard_block`
+
+```python
+def on_shard_block(store: Store, shard_store: ShardStore, signed_shard_block: SignedShardBlock) -> None:
+    shard_block = signed_shard_block.message
+    shard = shard_store.shard
+    # 1. Check shard parent exists
+    assert shard_block.shard_parent_root in shard_store.block_states
+    pre_shard_state = shard_store.block_states[shard_block.shard_parent_root]
+
+    # 2. Check beacon parent exists
+    assert shard_block.beacon_parent_root in store.block_states
+    beacon_state = store.block_states[shard_block.beacon_parent_root]
+
+    # 3. Check that block is later than the finalized epoch slot (optimization to reduce calls to get_ancestor)
+    finalized_slot = compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)
```

Comment:
I think this is a bit off

Instead, it is probably something like
* get the beacon state for the latest finalized
* get the shard state for the shard in the latest finalized
* ensure shard_block.slot is later than the shard state slot in the latest finalized beacon state (this can certainly be for a slot earlier than the latest finalized beacon slot